maven-model/src/main/java/org/apache/maven/model/merge/ModelMerger.java

```diff
@@ -3018,4 +2618,298 @@ protected Object getExclusionKey( Exclusion exclusion )
+        {
+            return getContributorKey( contributor );
+        }
+    }
+
+    class ProfileKeyComputer implements KeyComputer<Profile>
+    {
+        @Override
+        public Object key( Profile profile )
+        {
+            return getProfileKey( profile );
+        }
+    }
+
+    class RepositoryKeyComputer implements KeyComputer<Repository>
+    {
+        @Override
+        public Object key( Repository repository )
+        {
+            return getRepositoryKey( repository );
+        }
+    }
+
+    class ReportPluginKeyComputer implements KeyComputer<ReportPlugin>
+    {
+        @Override
+        public Object key( ReportPlugin plugin )
+        {
+            return getReportPluginKey( plugin );
+        }
+    }
+
+    class PluginKeyComputer implements KeyComputer<Plugin>
+    {
+        @Override
+        public Object key( Plugin plugin )
+        {
+            return getPluginKey( plugin );
+        }
+    }
+
+    class ReportSetKeyComputer implements KeyComputer<ReportSet>
+    {
+        @Override
+        public Object key( ReportSet reportSet )
+        {
+            return getReportSetKey( reportSet );
+        }
+    }
+
+    class NotifierKeyComputer implements KeyComputer<Notifier>
+    {
+        @Override
+        public Object key( Notifier notifier )
+        {
+            return getNotifierKey( notifier );
+        }
+    }
+
+    class ExtensionKeyComputer implements KeyComputer<Extension>
+    {
+        @Override
+        public Object key( Extension extension )
+        {
+            return getExtensionKey( extension );
+        }
+    }
+
+    class ResourceKeyComputer implements KeyComputer<Resource>
+    {
+        @Override
+        public Object key( Resource resource )
+        {
+            return getResourceKey( resource );
+        }
+    }
+
+    class ExecutionKeyComputer implements KeyComputer<PluginExecution>
+    {
+        @Override
+        public Object key( PluginExecution pluginExecution )
+        {
+            return getPluginExecutionKey( pluginExecution );
+        }
+    }
+
+    class ExclusionKeyComputer implements KeyComputer<Exclusion>
+    {
+        @Override
+        public Object key( Exclusion exclusion )
+        {
+            return getExclusionKey( exclusion );
+        }
+    }
+
+    static <T> List<T> merge( List<T> tgt, List<T> src, boolean sourceDominant, KeyComputer<T> computer )
+    {
+        if ( src.isEmpty() )
+        {
+            return tgt;
+        }
+
+        MergingList<T> list;
+        if ( tgt instanceof MergingList )
+        {
+            list = (MergingList<T>) tgt;
+        }
+        else
+        {
+            list = new MergingList<>( computer, src.size() + tgt.size() );
+            list.mergeAll( tgt, true );
+        }
+
+        list.mergeAll( src, sourceDominant );
+        return list;
+    }
+
+    /**
+     * Merging list
+     * @param <V>
+     */
+    public static class MergingList<V> extends AbstractList<V>
```

Comment:
@Tibor17 this is a brand new class, so there is no code that it is using it.
As @gnodet said we will make it private.

The only concern may be about returning a raw ArrayList instead of this special list after doing the merge.
It would be also interesting to make it unmodifiable but I don't know it we would break things.
Making it unmodifiable will make it threadsafe (the list, not the referenced objects)

@gnodet did this change in order to have better results, I guess this is not only a matter of refactoring to reduce code duplication

Reply:
the parent class is quite big but see in some parts this principle `c instanceof MerginList`.
The `synchronizedList` will break this condition. Not sure if Maven was designed wit memory safety. 
Maybe we can postpone the thoughts to the conference in Berlin and think about the improvements from high level perspectives. Personally I have to say that reciving and returing a Collection in a service method not like EE guys always do. I prefer a typical services where domain object is passed to a method and another is returned or void, but not Collection. The collection is structurally complext object and here maybe domain object with collection and value objects would need to be encapsulated. Then no `instanceof` is needed and instead an `isA` method would return true/false representing type of the collection upon the simple value object. Then we can talk about immutable and threadsafe and easily tested objects. Nobody would substitute the Collection with handsmade Collection and the `isA` method would work properly.

Reply:
@Tibor17 My goal here is just speed performances, nothing else. I don't want to refactor anything and my main goal is to keep 100% compatibility.  That's the reasons for the sometime weird choices that were made in this PR.
I don't think the `List` interface specific methods such as `listIterator`, `add(Object, int)` or such are used at all, but given the aim of 100% compatibility, I decided to make them available in the resulting collection.  So while your suggestions may be totally appropriate, this would not be compatible, so this is definitely a separate discussion, a separate jira issue, and a separate target release (i'm aiming for 100% compatibility to have this included in a 3.6.2 if possible).

The `c instanceof MergingList` has been added to be able to reuse this dual data structure and avoid switching between the map and list quite a few times.  I'll add a commit to make the `MergingList` private and the `KeyComputer` protected (so that they can be reused in the later derived classes).  Again the KeyComputer interface and implementations should completely go away once Maven requires JDK8, but again, that's a different target/discussion.